server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java

```diff
@@ -0,0 +1,188 @@
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.expression.tablefunctions;
+
+import io.crate.data.Input;
+import io.crate.data.Row;
+import io.crate.data.RowN;
+import io.crate.metadata.FunctionName;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.Signature;
+import io.crate.metadata.pgcatalog.PgCatalogSchemaInfo;
+import io.crate.metadata.tablefunctions.TableFunctionImplementation;
+import io.crate.types.DataTypes;
+import io.crate.types.RowType;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Locale;
+import java.util.NoSuchElementException;
+
+import static io.crate.metadata.functions.TypeVariableConstraint.typeVariable;
+import static io.crate.types.DataTypes.INTEGER;
+import static io.crate.types.TypeSignature.parseTypeSignature;
+
+public final class GenerateSubscripts<T> extends TableFunctionImplementation<T> {
+
+    public static final FunctionName NAME = new FunctionName(PgCatalogSchemaInfo.NAME, "generate_subscripts");
+    private static final RowType RETURN_TYPE = new RowType(List.of(INTEGER), List.of("subscripts"));
+
+
+    public static void register(TableFunctionModule module) {
+        module.register(
+            Signature.table(
+                NAME,
+                parseTypeSignature("array(E)"),
+                DataTypes.INTEGER.getTypeSignature(),
+                DataTypes.INTEGER.getTypeSignature()
+            ).withTypeVariableConstraints(typeVariable("E")),
+            GenerateSubscripts::new
+        );
+        module.register(
+            Signature.table(
+                NAME,
+                parseTypeSignature("array(E)"),
+                DataTypes.INTEGER.getTypeSignature(),
+                DataTypes.BOOLEAN.getTypeSignature(),
+                DataTypes.INTEGER.getTypeSignature()
+            ).withTypeVariableConstraints(typeVariable("E")),
+            GenerateSubscripts::new
+        );
+    }
+
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    private GenerateSubscripts(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    private static List<?> extractNestedArray(List<?> array, int targetLevel) {
+        List<?> targetArray = array;
+        for (int level = 2; level <= targetLevel; level++) {
+            if (targetArray == null || targetArray.isEmpty()) {
+                return null;
+            }
+            targetArray = checkAllElementsAreArraysOfConsistentDimension(targetArray, level);
+        }
+        return targetArray;
+    }
+
+    private static List<?> checkAllElementsAreArraysOfConsistentDimension(List<?> array, int level) {
```

Comment:
I think the name doesn't really fit. It implies that it verifies some condition, but also has a return value without indicating what the value is about.